mount/upgrader.go

```diff
@@ -96,11 +102,45 @@ func (u *Upgrader) Fetch(ctx context.Context) (Reader, error) {
 
 	once.Do(func() {
 		atomic.AddInt32(&u.fetches, 1)
-		err := u.refetch(ctx)
-		if err != nil {
-			log.Errorw("failed to refetch", "shard", u.key, "error", err)
+
+		// create a temporary file, partial file, and track it.
+		// onceErr can be set outside the lock because access is exclusive due to sync.Once
+		u.lk.Lock()
+		var file *os.File
+		file, u.onceErr = os.CreateTemp(u.rootdir, "transient-"+u.key+"-*-partial")
+		if u.onceErr != nil {
+			u.lk.Unlock()
+			return
+		}
+		defer file.Close()
+		u.partial = file.Name()
+		u.lk.Unlock()
+
+		// do the refetch; abort and remove/reset the partial if it fails.
+		u.onceErr = u.refetch(ctx, file)
+		if u.onceErr != nil {
+			log.Warnw("failed to refetch", "shard", u.key, "error", u.onceErr)
+			if err := os.Remove(u.partial); err != nil {
+				log.Warnw("failed to remove partial transient", "shard", u.key, "path", u.partial, "error", err)
+			}
+			u.partial = ""
```

Comment:
If the above lock is guarding the "partial" state, shouldn't we take a lock to mutate this ?